nomad/structs/acl.go

```diff
@@ -1156,6 +1207,117 @@ func (a *ACLAuthMethodConfig) UnmarshalJSON(data []byte) (err error) {
 }
 
+type OIDCClientAssertionKeySource string
+
+const (
+	OIDCKeySourceNomad        OIDCClientAssertionKeySource = "nomad"
+	OIDCKeySourceClientSecret OIDCClientAssertionKeySource = "client_secret"
+	OIDCKeySourcePrivateKey   OIDCClientAssertionKeySource = "private_key"
+)
+
+// OIDCClientAssertion (a.k.a private_key_jwt) is used to send
+// a client_assertion along with an OIDC token request.
+// See api.OIDCClientAssertion for full field descriptions.
+type OIDCClientAssertion struct {
+	KeySource    OIDCClientAssertionKeySource
+	Audience     []string
+	PrivateKey   *OIDCClientAssertionKey
+	ExtraHeaders map[string]string
+	KeyAlgorithm string
+	// clientSecret is inherited from parent ACLAuthMethodConfig struct
+	// via ACLAuthMethodConfig.Canonicalize
+	clientSecret string
+}
+
+func (c *OIDCClientAssertion) Canonicalize() {
+	if c == nil {
+		return
+	}
+	// default KeyAlgorithm to "RS256" for nomad and user keys, "HS256" for client_secret
+	if c.KeyAlgorithm == "" {
+		switch c.KeySource {
+		case OIDCKeySourceClientSecret:
+			c.KeyAlgorithm = "HS256"
+		case OIDCKeySourceNomad, OIDCKeySourcePrivateKey:
+			c.KeyAlgorithm = "RS256"
+		}
+	}
+}
+
+func (c *OIDCClientAssertion) IsSet() bool {
+	return c != nil && c.KeySource != ""
+}
+
+func (c *OIDCClientAssertion) ClientSecret() string {
+	return c.clientSecret
+}
+
+func (c *OIDCClientAssertion) Validate() error {
+	if c == nil { // guard, shouldn't happen
+		return errors.New("nil OIDCClientAssertion")
+	}
+	if len(c.Audience) == 0 {
+		return errors.New("audience is required")
+	}
+	switch c.KeySource {
+	case OIDCKeySourceNomad:
+	case OIDCKeySourcePrivateKey:
+		if c.PrivateKey == nil {
+			return errors.New("PrivateKey is required for `private_key` key source")
+		}
+		if err := c.PrivateKey.Validate(); err != nil {
+			return fmt.Errorf("invalid PrivateKey: %w", err)
+		}
+	case OIDCKeySourceClientSecret:
+		if c.clientSecret == "" {
+			return errors.New("OIDCClientSecret is required for `client_secret` key source")
+		}
+	default:
+		return fmt.Errorf("invalid KeySource %q", c.KeySource)
+	}
+	return nil
+}
+
+// OIDCClientAssertionKey contains key material provided by users for Nomad
+// to use to sign the private key JWT.
+// See api.OIDCClientAssertionKey for full field descriptions.
+type OIDCClientAssertionKey struct {
+	Base64PemKey string
+	PemKeyFile   string
+
+	Base64PemCert string
+	PemCertFile   string
+	KeyID         string
+}
+
+// Validate ensures that one Key and one Cert or KeyID are provided.
+func (k *OIDCClientAssertionKey) Validate() error {
+	if k == nil { // guard, shouldn't happen
+		return errors.New("nil OIDCClientAssertionKey")
+	}
+	// mutual exclusive fields
+	me := set.From([]string{k.Base64PemKey, k.PemKeyFile})
+	if !me.Remove("") { // one should be empty
+		return errors.New("require exactly one of Base64PemKey or PemKeyFile")
+	}
+	if me.Size() == 0 { // but not both of them
+		return errors.New("missing Base64PemKey or PemKeyFile")
+	}
```

Comment:
Why do all this memory allocation instead of a couple of conditionals?

```suggestion
	// mutual exclusive fields
	if k.Base64PemKey != "" && k.PemKeyFile != "" {
		return errors.New("require exactly one of Base64PemKey or PemKeyFile")
	}
	if k.Base64PemKey == "" && k.PemKeyFile == "" {	
		return errors.New("missing Base64PemKey or PemKeyFile")
	}
```

Reply:
I did this straightforward way initially, which was easy for 2 things, but it got unwieldy with 3 things (cert key/file/keyid), so I got clever with that, then did the same "clever" thing here for consistency.

I will un-clever the both of them!